front/lib/api/assistant/generation.ts

```diff
@@ -70,46 +72,91 @@ export async function renderConversationForModelMultiActions({
       const actions = removeNulls(m.actions);
 
       // This array is 2D, because we can have multiple calls per agent message (parallel calls).
-      const steps = [] as Array<
-        Array<{ call: FunctionCallType; result: FunctionMessageTypeModel }>
-      >;
+
+      const steps = [] as Array<{
+        contents: string[];
+        actions: Array<{
+          call: FunctionCallType;
+          result: FunctionMessageTypeModel;
+        }>;
+      }>;
+
+      const emptyStep = () =>
+        ({
+          contents: [],
+          actions: [],
+        }) satisfies (typeof steps)[number];
 
       if (!excludeActions) {
         for (const action of actions) {
           const stepIndex = action.step;
-          steps[stepIndex] = steps[stepIndex] || [];
-          steps[stepIndex].push({
+          steps[stepIndex] = steps[stepIndex] || emptyStep();
+          steps[stepIndex].actions.push({
             call: action.renderForFunctionCall(),
             result: action.renderForMultiActionsModel(),
           });
         }
       }
 
-      for (const step of steps) {
-        if (!step?.length) {
-          logger.error(
-            {
-              workspaceId: conversation.owner.sId,
-              conversationId: conversation.sId,
-              agentMessageId: m.sId,
-              panic: true,
-            },
-            "Unexpected state, agent message step with no actions"
-          );
-          continue;
+      for (const content of m.rawContents) {
+        steps[content.step] = steps[content.step] || emptyStep();
+        if (content.content.trim()) {
+          steps[content.step].contents.push(content.content);
         }
-        messages.push({
-          role: "assistant",
-          function_calls: step.map((s) => s.call),
-        });
-        for (const { result } of step) {
-          messages.push(result);
+      }
+
+      if (excludeActions) {
```

Comment:
wait if excludeActions is true then steps is []?

Reply:
I see this is quite hard to reason about... can we instead compute `generationStep` once and then use that to inject the generation contents whether excludeActions is true or not ?

Reply:
We can also use that to only add content to the generationStep if excludeAction is true in the loop above.

From there the code below becomes much simpler?

Reply:
Calculating `generationStep` we can have safe-guard checking that there is no tool use at this step etc to make sure we're in a good state